p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,71 @@ extern Checksum<W> {
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+'indirect' counters, which are implemented by the extern `Counter`.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
```

Comment:
As mentioned in the comment to another commit, the semantics of DirectCounter used together with ActionProfile needs to be clarified. Here are a couple of options:
1. DirectCounter cannot be used on tables that use `implementation` property
2. Specifically for tables with ActionProfiles, the number of direct counters is equal to the number of entries in the action profile
3. Regardless of the implementation, the number of counters is equal to the number of match entries

The last option is the cleanest on the surface, but since counting is an action, it might be difficult to implement on all targets